File: jscomp/test/format_regression.js
'use strict';

var Curry = require("../../lib/js/curry.js");
var Caml_builtin_exceptions = require("../../lib/js/caml_builtin_exceptions.js");

function peek_queue(param) {
  throw [
        Caml_builtin_exceptions.assert_failure,
        /* tuple */[
          "format_regression.ml",
          10,
          19
        ]
      ];
}

function int_of_size(param) {
  throw [
        Caml_builtin_exceptions.assert_failure,
        /* tuple */[
          "format_regression.ml",
          11,
          20
        ]
      ];
}

function take_queue(param) {
  throw [
        Caml_builtin_exceptions.assert_failure,
        /* tuple */[
          "format_regression.ml",
          12,
          19
        ]
      ];
}

function format_pp_token(param, param$1) {
  throw [
        Caml_builtin_exceptions.assert_failure,
        /* tuple */[
          "format_regression.ml",
          13,
          26
        ]
      ];
}

function advance_loop(state) {
  while(true) {
    var match = peek_queue(state.pp_queue);
    var size = match.elem_size;
    var size$1 = int_of_size(size);
    if (size$1 < 0 && (state.pp_right_total - state.pp_left_total | 0) < state.pp_space_left) {
      return 0;
    } else {
      take_queue(state.pp_queue);
      Curry._1(format_pp_token(state, size$1 < 0 ? 1000000010 : size$1), match.token);
      state.pp_left_total = match.length + state.pp_left_total | 0;
      continue ;
    }
  };
}

var pp_infinity = 1000000010;

exports.peek_queue = peek_queue;
exports.int_of_size = int_of_size;
exports.take_queue = take_queue;
exports.format_pp_token = format_pp_token;
exports.pp_infinity = pp_infinity;
exports.advance_loop = advance_loop;
/* No side effect */

File: jscomp/test/format_regression.ml
[@@@warning "-20-21"]

type tag
type 'a queue_elem
type 'a queue = {mutable insert: 'a queue_elem; mutable body: 'a queue_elem}
type size
type pp_token

let peek_queue _ = assert false
let int_of_size _ = assert false
let take_queue _ = assert false
let format_pp_token _ _ = assert false
let pp_infinity = 1000000010

type pp_queue_elem = {mutable elem_size: size; token: pp_token; length: int}

type formatter =
  { mutable pp_space_left: int
  ; mutable pp_left_total: int
  ; mutable pp_right_total: int
  ; mutable pp_queue: pp_queue_elem queue }

let rec advance_loop state =
  match peek_queue state.pp_queue with
  | {elem_size= size; token= tok; length= len} ->
      let size = int_of_size size in
      if
        not
          ( size < 0
          && state.pp_right_total - state.pp_left_total < state.pp_space_left
          )
      then (
        ignore (take_queue state.pp_queue) ;
        format_pp_token state (if size < 0 then pp_infinity else size) tok ;
        state.pp_left_total <- len + state.pp_left_total ;
        advance_loop state )
